src/lib/custom-resources.ts

```diff
@@ -0,0 +1,741 @@
+		} else {
+			fileList.push(path.resolve(dir, file));
+		}
+	}
+
+	return fileList;
+}
+
+function getFrame(src: Canvas, frame: number) {
+	const dst = createCanvas(normalizedSize, normalizedSize);
+	const ctx = dst.getContext('2d');
+
+	ctx.drawImage(src, 0, frame * normalizedSize * -1);
+
+	return dst;
+}
+
+let resourcePacks: CustomResourcesResourcePack[] = [];
+let packConfigHashes: { [key: string]: string } = {};
+
+const outputPacks: CustomResourcesOutputPack[] = [];
+
+export async function init() {
+	const timeNow = performance.now();
+	console.log(`Custom Resources loading started.`);
+
+	await loadPackConfigs();
+	let resourcesUpToDate = false;
+
+	try {
+		packConfigHashes = {};
+		packConfigHashes = JSON.parse(fs.readFileSync(packHashCacheFile, 'utf8'));
+
+		if (Object.keys(packConfigHashes).length !== resourcePacks.length) {
+			throw new Error('The amount of config hashes does not match the amount of packs');
+		}
+
+		resourcePacks.forEach((pack) => {
+			if (packConfigHashes[pack.config.id] !== pack.config.hash) {
+				throw new Error('Config hashes were not matching!');
+			}
+		});
+
+		resourcesUpToDate = true;
+	} catch (e) {
+		packConfigHashes = {};
+		resourcePacks.forEach((pack: CustomResourcesResourcePack) => {
+			packConfigHashes[pack.config.id] = pack.config.hash;
+		});
+
+		fs.writeFileSync(packHashCacheFile, JSON.stringify(packConfigHashes));
+	}
+
+	try {
+		if (!resourceCaching) {
+			throw new Error('Resource caching has been disabled!');
+		}
+
+		if (resourcesUpToDate) {
+			resourcePacks = JSON.parse(fs.readFileSync(resourceCacheFile, 'utf8'));
+		} else {
+			throw new Error('Resources need to be loaded!');
+		}
+	} catch (e) {
+		await loadResourcePacks();
+
+		fs.writeFileSync(resourceCacheFile, JSON.stringify(resourcePacks));
+	}
+
+	resourcePacks = resourcePacks.sort((a, b) => b.config.priority - a.config.priority);
+	resourcePacks.forEach((pack) => {
+		outputPacks.push(
+			Object.assign(
+				{
+					base_path: '/' + path.relative(path.resolve(folderPath, '..', 'public'), pack.base_path)
+				},
+				pack.config
+			)
+		);
+	});
+
+	resourcesReady = true;
+
+	console.log(`Custom Resources loading done in ${(performance.now() - timeNow).toFixed(2).toLocaleString()}ms.`);
+}
+
+async function loadPackConfigs() {
+	for (const packOrFile of await fs.readdir(resourcePackFolder, { withFileTypes: true })) {
+		if (!packOrFile.isDirectory()) {
+			continue;
+		}
+
+		const pack = packOrFile.name;
+		const basePath = path.resolve(resourcePackFolder, pack);
+
+		try {
+			const configPath = path.resolve(basePath, 'config.json');
+
+			const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
+			config.hash = await getFileHash(configPath);
+
+			resourcePacks.push({
+				base_path: basePath,
+				config,
+				textures: []
+			});
+		} catch (e) {
+			console.log("Couldn't find config for resource pack", pack);
+		}
+	}
+}
+
+async function loadResourcePacks() {
```

Comment:
It's up to you because you're likely the only one that's going to maintain this function, but this is _massive_. It'd be way more readable/maintainable if these inner branches would be extracted into their own functions.

Reply:
I've got plans to rewrite it, but it's not happening anytime soon cuz It's a bit complex, so I'll probably get to it in about 2-3 weeks when I'm gonna actually be free. I just wanted to get it working so @tonydawhale can start on the front end